src/Web/Internal/FormUrlEncoded.hs

```diff
@@ -0,0 +1,499 @@
+
+instance (FromFormKey k, FromHttpApiData v) => FromForm [(k, v)] where
+  fromForm = fmap (concatMap (\(k, vs) -> map ((,) k) vs)) . toEntriesByKey
+
+instance (Ord k, FromFormKey k, FromHttpApiData v) => FromForm (Map k [v]) where
+  fromForm = fmap (Map.fromListWith (<>)) . toEntriesByKey
+
+instance (Eq k, Hashable k, FromFormKey k, FromHttpApiData v) => FromForm (HashMap k [v]) where
+  fromForm = fmap (HashMap.fromListWith (<>)) . toEntriesByKey
+
+instance FromHttpApiData v => FromForm (IntMap [v]) where
+  fromForm = fmap (IntMap.fromListWith (<>)) . toEntriesByKey
+
+-- | Parse a 'Form' into a list of entries groupped by key.
+toEntriesByKey :: (FromFormKey k, FromHttpApiData v) => Form -> Either Text [(k, [v])]
+toEntriesByKey = traverse parseGroup . HashMap.toList . unForm
+  where
+    parseGroup (k, vs) = (,) <$> parseFormKey k <*> traverse parseQueryParam vs
+
+-- | A 'Generic'-based implementation of 'fromForm'.
+-- This is used as a default implementation in 'FromForm'.
+genericFromForm :: (Generic a, GFromForm (Rep a))
+    => Form -> Either Text a
+genericFromForm f = to <$> gFromForm f
+
+class GFromForm (f :: * -> *) where
+  gFromForm :: Form -> Either Text (f x)
+
+instance (GFromForm f, GFromForm g) => GFromForm (f :*: g) where
+  gFromForm f = (:*:) <$> gFromForm f <*> gFromForm f
+
+instance (GFromForm f, GFromForm g) => GFromForm (f :+: g) where
+  gFromForm f
+      = fmap L1 (gFromForm f)
+    <!> fmap R1 (gFromForm f)
+    where
+      Left _  <!> y = y
+      x       <!> _ = x
+
+instance (Selector s, FromHttpApiData f) => GFromForm (M1 S s (K1 i f)) where
+  gFromForm form = M1 . K1 <$> parseUnique key form
+    where
+      key = Text.pack $ selName (Proxy3 :: Proxy3 s g p)
+
+instance (GFromForm f) => GFromForm (M1 D x f) where
+  gFromForm f = M1 <$> gFromForm f
+
+instance (GFromForm f) => GFromForm (M1 C x f) where
+  gFromForm f = M1 <$> gFromForm f
+
+-- | Encode a 'Form' to an @application/x-www-form-urlencoded@ 'BSL.ByteString'.
+--
+-- Key-value pairs get encoded to @key=value@ and separated by @&@:
+--
+-- >>> encodeForm [("name", "Julian"), ("lastname", "Arni")]
+-- "lastname=Arni&name=Julian"
+--
+-- Keys with empty values get encoded to just @key@ (without the @=@ sign):
+--
+-- >>> encodeForm [("is_test", "")]
+-- "is_test"
+--
+-- Empty keys are allowed too:
+--
+-- >>> encodeForm [("", "foobar")]
+-- "=foobar"
+--
+-- However, if not key and value are empty, the key-value pair is ignored.
+-- (This prevents @'decodeForm' . 'encodeForm'@ from being a true isomorphism).
+--
+-- >>> encodeForm [("", "")]
+-- ""
+--
+-- Everything is escaped with @'escapeURIString' 'isUnreserved'@:
+--
+-- >>> encodeForm [("fullname", "Andres Löh")]
+-- "fullname=Andres%20L%C3%B6h"
+encodeForm :: Form -> BSL.ByteString
+encodeForm xs = BSL.intercalate "&" $ map (BSL.fromStrict . encodePair) $ toList xs
+  where
+    escape = Text.encodeUtf8 . Text.pack . escapeURIString isUnreserved . Text.unpack
```

Comment:
This is obviously inefficient. I'm in favor of making changes incrementally, and so leaving the fixing of it for another PR, but for reference [`uri-bytestring`s]((https://hackage.haskell.org/package/uri-bytestring-0.2.2.0/docs/URI-ByteString.html) could ) `urlEncode` could be a good option.